File: axion/AxionTokens.py
# author https://github.com/MIrrox27/Axion-Language
# AxionTokens.py

from enum import Enum


# Перечисление всех типов токенов в нашем языке программирования
class AxionTokenType(Enum):
    # ЛИТЕРАЛЫ (константные значения)
    INTEGER = 'INTEGER'  # Целое число: 42, 100, -5
    STRING = 'STRING'  # Строка в кавычках: "привет", "мир"
    FLOAT = 'FLOAT'  # Десятичное число: 3.14, 2.5, 0.5
    BOOL = 'BOOL'  # Логическое значение: true, false
    FUNCTION_TYPE = 'FUNCTION_TYPE' # Особый тип данных для использования в словарях и списках (понадобится в будущем)
    ALL = 'ALL' # Специальный тип данных для использования в словарях и списках, показывает что можно использовать любой тип данных указанных выше (понадобится в будущем)


    # КЛЮЧЕВЫЕ СЛОВА (зарезервированные слова языка)
    FUN = 'FUN'  # Объявление функции: fun(value){}
    CONFIG = 'CONFIG' # Объявление конфига: config{} (для использования в линуксе)
    SCRIPT = 'SCRIPT' # Объявление начала скрипта: script{} (для использования в линуксе)
    CLASS = 'CLASS'  # Объявление класса: class(наследование){}
    ENUM = 'ENUM'  # Объявление перечисления: enum{}
    IMPORT = 'IMPORT'  # Импорт модулей: import module
    VAR = 'VAR'  # Объявление изменяемой переменной var __name__ = value <- not const
    VAL = 'VAL' # Объявление неизменяемой переменной val __name__ = value <- const
    IF = 'IF'  # Условный оператор: if(parameters){}
    ELIF = 'ELIF' # Условный оператор: elif(parameters){}
    ELSE = 'ELSE'  # Условный оператор: else{}
    WHILE = 'WHILE'  # Цикл while: while
    DO = 'DO' # Цикл do (образован от цикла do while), первый раз выполняется всегда, потом по условию
    FOR = "FOR"  # Цикл for: for
    FOREACH = "FOREACH" # Цикл foreach: foreach(variable in collection)
    RETURN = 'RETURN' # Возврат значения: return
    IS = 'IS' # Ключевое слово is
    IN = 'IN' # Ключевое слово in
    AND = 'AND' # Ключевое слово and (&&)
    OR = 'OR' # Ключевое слово or (||)
    NONE = 'NONE' # Ключевое слово nill (мб, сделаем так, чтобы можно было заменить на ";;")
    NOT = 'NOT' # Ключевое слово not, так же может обозначаться знаком "!"




    BLOCK = 'BLOCK'  # Ключевое слово для обозначения исполнения блока кода на другом языке block __lang__{axion}
    PYTHON = 'PYTHON'  # Ключевое слово для исполнения кода на питоне

    CODE_BLOCK =  'CODE_OPEN' # сочетание символов внутри которых находится код на другом языке: {$ __code__ $}
    #CODE_CLOSE = 'CODE_CLOSE' # символ закрывающий блок кода: $}

    ARROW = 'ARROW' # символ -> (для указания возвращаемых переменных)


    # ОПЕРАТОРЫ (математические и логические операции)
    PLUS = 'PLUS'  # Сложение: +
    MINUS = 'MINUS'  # Вычитание: -
    INCREMENT = 'INCREMENT' # ++x
    DECREMENT = 'DECREMENT' # --x
    POST_INCREMENT = 'POST_INCREMENT'  # x++
    POST_DECREMENT = 'POST_DECREMENT'  # x--
    MULTIPLY = 'MULTIPLY'  # Умножение: *
    DIVIDE = 'DIVIDE'  # Деление: /
    ASSIGN = 'ASSIGN'  # Присваивание: =
    EQUALS = 'EQUALS'  # Равенство: ==
    NOT_EQUALS = 'NOT_EQUALS'  # Неравенство: !=
    LESS = 'LESS'  # Меньше: <
    GREATER = 'GREATER'  # Больше: >
    MOD = 'MOD' # %
    POWER = 'POWER' # степень числа: **
    LESS_EQUAL = 'LESS_EQUAL' # <=
    GREATER_EQUAL = 'GREATER_EQUAL' # >=

    # РАЗДЕЛИТЕЛИ (пунктуация и скобки)
    LPAREN = 'LPAREN'  # Левая круглая скобка: (
    RPAREN = 'RPAREN'  # Правая круглая скобка: )
    LBRACE = 'LBRACE'  # Левая фигурная скобка: {
    RBRACE = 'RBRACE'  # Правая фигурная скобка: }
    LBRACKET = 'LBRACKET' # Левая квадратная скобка: [
    RBRACKET = 'RBRACKET' # Правая квадратная скобка: ]
    SEMICOLON = 'SEMICOLON'  # Точка с запятой: ;
    COLON = 'COLON' # Двоеточие: :
    COMMA = 'COMMA'  # Запятая: ,
    DOT = 'DOT' # точка .

    # СПЕЦИАЛЬНЫЕ ТОКЕНЫ
    EOF = 'EOF'  # Конец файла (End Of File)
    IDENTIFIER = 'IDENTIFIER'  # Идентификатор (имя переменной/функции)

    AXION = "AXION"  # Проверочный токен для языка Axion


# Класс для представления отдельного токена
class AxionToken:
    def __init__(self, type: AxionTokenType, value: any = None, line: int = 1):
        import builtins
        if not getattr(builtins, '__axion_verified__', False):
            raise RuntimeError("Unauthorized copy of Axion interpreter detected.")

        self.type = type  # Тип токена из перечисления TokenType
        self.value = value  # Значение токена (для чисел, строк, идентификаторов)
        self.line = line  # Номер строки, где найден токен (для отладки ошибок)

    def __str__(self):
        # Красивое строковое представление токена для отладки
        if self.value is not None:
            return f"Token({self.type}, {repr(self.value)}, line: {self.line})"
        return f"Token({self.type}, line: {self.line})"

    def __repr__(self):
        # Техническое строковое представление
        return self.__str__()

File: src/axion/axion_tokens.rs
// author https://github.com/MIrrox27/Axion-Language

#![allow(dead_code)]
#[derive(Debug)]

pub enum AxionTokenType{
  IntegerType,  // Целое Число: 42, 100, -5
  StringType,  // Строка В Кавычках: Привет, Мир
  FloatType,  // Десятичное Число: 3.14, 2.5, 0.5
  BoolType,  // Логическое Значение: True, False
  FunctionType,  // Особый Тип Данных Для Использования В Словарях И Списках (понадобится В Будущем)
  AllType, // Специальный Тип Данных Для Использования В Словарях И Списках, Показывает Что Можно Использовать Любой Тип Данных Указанных Выше (понадобится В Будущем)


   // Ключевые Слова (зарезервированные Слова Языка)
  Fun,  // Объявление Функции: Fun(value){}
  Config,  // Объявление Конфига: Config{} (для Использования В Линуксе)
  Script,  // Объявление Начала Скрипта: Script{} (для Использования В Линуксе)
  Class,  // Объявление Класса: Class(наследование){}
  Enum,  // Объявление Перечисления: Enum{}
  Import,  // Импорт Модулей: Import Module
  Var,  // Объявление Изменяемой Переменной Var __name__  Value <- Not Const
  Val,  // Объявление Неизменяемой Переменной Val __name__  Value <- Const
  If,  // Условный Оператор: If(parameters){}
  Elif,  // Условный Оператор: Elif(parameters){}
  Else,  // Условный Оператор: Else{}
  While,  // Цикл While: While
  Do,  // Цикл Do (образован От Цикла Do While), Первый Раз Выполняется Всегда, Потом По Условию
  For,  // Цикл For: For
  Foreach,  // Цикл Foreach: Foreach(variable In Collection)
  Return,  // Возврат Значения: Return
  Is,  // Ключевое Слово Is
  In,  // Ключевое Слово In
  And,  // Ключевое Слово And (&&)
  Or,  // Ключевое Слово Or (||)
  None, // Ключевое Слово Nill (мб, Сделаем Так, Чтобы Можно Было Заменить На ;;)
  Not,  // Ключевое Слово Not, Так Же Может Обозначаться Знаком !




  Block,  // Ключевое Слово Для Обозначения Исполнения Блока Кода На Другом Языке Block __lang__{axion}
  Python,  // Ключевое Слово Для Исполнения Кода На Питоне
  CodeOpen, // Сочетание Символов Внутри Которых Находится Код На Другом Языке: {$ __code__ $}
  //code_close  Code_close // Символ Закрывающий Блок Кода: $}
  Arrow, // Символ -> (для Указания Возвращаемых Переменных)


    // Операторы (математические И Логические Операции)
  Plus,  // Сложение: +
  Minus,  // Вычитание: -
  Increment,  // ++x
  Decrement,  // --x
  PostIncrement,  // X++
  PostDecrement,  // X--
  Multiply,  // Умножение: *
  Divide,  // Деление: /
  Assign,  // Присваивание: 
  Equals,  // Равенство: 
  NotEquals,  // Неравенство: !
  Less,  // Меньше: <
  Greater,  // Больше: >
  Mod,  // %
  Power, // Степень Числа: **
  LessEqual,  // <
  GreaterEqual, // >


      // Разделители (пунктуация И Скобки)
  LParen,  // Левая Круглая Скобка: (
  RParen,  // Правая Круглая Скобка: )
  LBrace,  // Левая Фигурная Скобка: {
  RBrace,  // Правая Фигурная Скобка: }
  LBracket, // Левая Квадратная Скобка: [
  RBracket,  // Правая Квадратная Скобка: ]
  Semicolon,  // Точка С Запятой: ;
  Colon,  // Двоеточие: :
  Comma,  // Запятая: ,
  Dot,  // Точка .


  // Специальные Токены
  Eof,  // Конец Файла (end Of File)
  Identifier,  // Идентификатор (имя Переменной/функции)


  Axion,  // Проверочный Токен Для Языка Axion
}






#[derive(Debug)]
pub struct AxionToken {
    pub token_type: AxionTokenType,
    pub value: Option<String>,
    pub line: usize

} 

impl AxionToken {
    pub fn new(token_type: AxionTokenType, value: Option<String>, line: usize) -> Self{
      AxionToken { token_type, value, line }
    }
}
